### PR TITLE
feat: add enrollmentv2

### PR DESCRIPTION
**Pull request by Ali-D-Akbar**
[PROD-2462](https://openedx.atlassian.net/browse/PROD-2462)
This PR adds does the following while keeping the previous components:
1. A new route `/usersv2` for the redesigned components
2. A custom table component `TableV2` as per the requirements in the ticket which can take an expandable cell and allows custom column based sorting
3. A new Enrollment component `EnrollmentV2` and it's unit tests

These components will eventually replace the older components. Below are some screenshots for `EnrollmentV2`

2 Enrollments
Screenshot 2021-09-03 at 5 13 48 PM

0 Enrollments
Screenshot 2021-09-03 at 5 16 39 PM

Upon clicking `+` icon
Screenshot 2021-09-03 at 5 22 58 PM

Upon clicking `expand all`
Screenshot 2021-09-03 at 5 24 00 PM

Upon loading
Screenshot 2021-09-03 at 5 24 23 PM

### CODE REVIEW

**Comment by azanbinzahid** [comment]
Can we somehow avoid userpagev2 by rendering tablev2 if route has v2 in it?

**Comment by Ali-D-Akbar** [comment]
> Can we somehow avoid userpagev2 by rendering tablev2 if route has v2 in it?

@azanbinzahid we're gonna have a lot more than just a table. Since we'll be doing a complete re-design, I think this will be a better approach for the future when we want to add more v2 components.

**Comment by azanbinzahid** [comment]
Make sense, what about having a `v2` sub-directory in each component or in `src` and routing all traffic via `users/v2`. It will be much cleaner once we scale.

File: src/TableV2.jsx [hunk 1]
```diff
@@ -0,0 +1,99 @@
+import React from 'react';
```
Comment (DawoudSheraz) [inline src/TableV2.jsx]: moce this to components/Table/Tablev2.jsx

File: src/index.jsx [hunk 2]
```diff
@@ -35,6 +36,7 @@ subscribe(APP_READY, () => {
         <Switch>
           <Route exact path="/" component={SupportHomePage} />
           <Route path="/users" component={UserPage} />
+          <Route path="/usersv2" component={UserPageV2} />
```
Comment (DawoudSheraz) [inline src/index.jsx]: I would say avoid adding this route for now. Once we have enough new components, only then we should add a new route. In case of a support team member reaching this new page by accident, it will cause confusion on their end.

File: src/users/enrollments/EnrollmentsV2.jsx [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+  });
+
+  const expandAllHandler = ({ getToggleAllRowsExpandedProps }) => (
+    <a {...getToggleAllRowsExpandedProps()} className="link-primary">
+      Expand All
+    </a>
+  );
+  expandAllHandler.propTypes = {
+    getToggleAllRowsExpandedProps: PropTypes.func.isRequired,
+  };
```
Comment (DawoudSheraz) [inline src/users/enrollments/EnrollmentsV2.jsx]: why is this added here?

File: src/users/enrollments/EnrollmentsV2.jsx [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+      </span>
+    </div>
+  );
+
+  expandHandler.propTypes = {
+    row: PropTypes.shape({
+      isExpanded: PropTypes.bool,
+      getToggleRowExpandedProps: PropTypes.func,
+    }).isRequired,
+  };
```
Comment (DawoudSheraz) [inline src/users/enrollments/EnrollmentsV2.jsx]: same question here

File: src/users/enrollments/EnrollmentsV2.jsx [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+        Header: 'Active', accessor: 'active', sortable: true,
+      },
+      {
+        Header: 'Actions', accessor: 'actions',
+      },
+    ],
+    [],
+  );
+
+  const extraColumns = React.useMemo(
```
Comment (DawoudSheraz) [inline src/users/enrollments/EnrollmentsV2.jsx]: what benefit does useMemo provide here?

File: src/users/enrollments/EnrollmentsV2.jsx [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+  const expandAllHandler = ({ getToggleAllRowsExpandedProps }) => (
+    <a {...getToggleAllRowsExpandedProps()} className="link-primary">
+      Expand All
+    </a>
+  );
+  expandAllHandler.propTypes = {
+    getToggleAllRowsExpandedProps: PropTypes.func.isRequired,
+  };
+
+  const expandHandler = ({ row }) => (
```
Comment (DawoudSheraz) [inline src/users/enrollments/EnrollmentsV2.jsx]: nit: rowExpandHandler

File: src/users/enrollments/EnrollmentsV2.jsx [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+    }));
+  }, [enrollmentData]);
+
+  useLayoutEffect(() => {
+    if (formType != null) {
+      formRef.current.focus();
+    }
+  });
+
+  const expandAllHandler = ({ getToggleAllRowsExpandedProps }) => (
```
Comment (DawoudSheraz) [inline src/users/enrollments/EnrollmentsV2.jsx]: nit: expandAllRowsHandler

File: src/users/enrollments/EnrollmentsV2.jsx [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+        Header: 'Last Modified By', accessor: 'lastModifiedBy',
+      },
+      {
+        Header: 'Reason', accessor: 'reason',
+      },
+    ],
+    [],
+  );
+
+  const renderRowSubComponent = useCallback(
```
Comment (DawoudSheraz) [inline src/users/enrollments/EnrollmentsV2.jsx]: why use callback?

File: src/users/enrollments/EnrollmentsV2.jsx [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+  );
+
+  expandHandler.propTypes = {
+    row: PropTypes.shape({
+      isExpanded: PropTypes.bool,
+      getToggleRowExpandedProps: PropTypes.func,
+    }).isRequired,
+  };
+
+  const columns = React.useMemo(
```
Comment (DawoudSheraz) [inline src/users/enrollments/EnrollmentsV2.jsx]: how does use memo benefit here?

File: src/users/enrollments/EnrollmentsV2.jsx (hunk 7 above)
Comment (DawoudSheraz) [reply src/users/enrollments/EnrollmentsV2.jsx]: getToggleAllRowsExpandedProps --> is this provided internally by react-table?

File: src/users/enrollments/EnrollmentsV2.jsx [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+  );
+  expandAllHandler.propTypes = {
+    getToggleAllRowsExpandedProps: PropTypes.func.isRequired,
+  };
+
+  const expandHandler = ({ row }) => (
+    // We can use the getToggleRowExpandedProps prop-getter
+    // to build the expander.
+    <div className="text-center">
+      <span {...row.getToggleRowExpandedProps()}>
```
Comment (DawoudSheraz) [inline src/users/enrollments/EnrollmentsV2.jsx]: what does this return?

File: src/users/enrollments/EnrollmentsV2.test.jsx [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+    jest.spyOn(api, 'getEnrollments').mockImplementationOnce(() => Promise.resolve(enrollmentsData));
+    wrapper = mount(<EnrollmentPageWrapper {...props} />);
+    await waitForComponentToPaint(wrapper);
+  });
+
+  afterEach(() => {
+    wrapper.unmount();
+  });
+
+  it('default collapsible with enrollment data', () => {
```
Comment (DawoudSheraz) [inline src/users/enrollments/EnrollmentsV2.test.jsx]: The new enrollment table is not using collapsible. Please update the text.

File: src/users/enrollments/EnrollmentsV2.test.jsx [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+    expect(extraTableHeaders.at(1).text()).toEqual('Last Modified By');
+    expect(extraTableHeaders.at(2).text()).toEqual('Reason');
+
+    expandable.simulate('click');
+
+    expandable = wrapper.find('table tbody tr').at(0).find('td div span').at(0);
+    expect(expandable.html()).toContain('fa-plus');
+  });
+
+  it('Expand all button shows extra data for all enrollments', () => {
```
Comment (DawoudSheraz) [inline src/users/enrollments/EnrollmentsV2.test.jsx]: add a test to verify no action takes place when Expand all is clicked with no table data.

File: src/users/enrollments/EnrollmentsV2.test.jsx [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+    await waitForComponentToPaint(wrapper);
+
+    const alert = wrapper.find('div.alert');
+    expect(alert.text()).toEqual(enrollmentErrors.errors[0].text);
+  });
+
+  it('Enrollment create form is rendered', () => {
+    const createEnrollmentButton = wrapper.find('button#create-enrollment-button');
+    createEnrollmentButton.simulate('click');
+    const createEnrollmentForm = wrapper.find('CreateEnrollmentForm');
```
Comment (DawoudSheraz) [inline src/users/enrollments/EnrollmentsV2.test.jsx]: Verify that enrollmentform has been rendered. The current test is not verifying the existence of the form.

**Review by DawoudSheraz — COMMENTED** [review]
bunch of comments to address and questions to answer

File: src/index.jsx (hunk 2 above)
Comment (azanbinzahid) [reply src/index.jsx]: We may:
- use `users/v2/` as parent route (and change history pushes accordingly)
- comment this line or enable this only using an environment variable locally

File: src/users/enrollments/EnrollmentsV2.test.jsx [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+    <EnrollmentsV2 {...props} />
+  </UserMessagesProvider>
+);
+
+describe('Course Enrollments V2 Listing', () => {
+  let wrapper;
+  const props = {
+    user: 'edX',
+    changeHandler: jest.fn(() => {}),
+    expanded: true,
```
Comment (azanbinzahid) [inline src/users/enrollments/EnrollmentsV2.test.jsx]: I guess we no longer need `expanded` prop.

File: src/TableV2.jsx [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+import { useTable, useExpanded, useSortBy } from 'react-table';
+import PropTypes from 'prop-types';
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+import { faSortDown, faSortUp } from '@fortawesome/free-solid-svg-icons';
+
+export default function TableV2({
```
Comment (azanbinzahid) [inline src/TableV2.jsx]: minor: just for the consistency, this can remain as `Table` or add V2 to `UserPage` and `Enrollments` before exporting.

File: src/overrides.scss [hunk 16]
```diff
@@ -1,6 +1,42 @@
...
+    margin: 0;
+    padding: 0.5rem;
+    border-bottom: 1px solid black;
+  }
+  td {
+    margin: 0;
+    padding: 0.5rem;
+    border-bottom: 1px solid silver;
+  }
+}
```
Comment (azanbinzahid) [inline src/overrides.scss:22]: will this scale well with changing screen sizes?

File: src/users/enrollments/EnrollmentsV2.jsx (hunk 3 above)
Comment (Ali-D-Akbar) [reply src/users/enrollments/EnrollmentsV2.jsx]: Each function must have prop validation: https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/prop-types.md

File: src/users/enrollments/EnrollmentsV2.jsx (hunk 5 above)
Comment (Ali-D-Akbar) [reply src/users/enrollments/EnrollmentsV2.jsx]: react-table component uses memoized data: https://react-table.tanstack.com/docs/quick-start#define-columns

File: src/users/enrollments/EnrollmentsV2.jsx (hunk 7 above)
Comment (Ali-D-Akbar) [reply src/users/enrollments/EnrollmentsV2.jsx]: Yes, table sends this function `getToggleAllRowsExpandedProps ` into `expandAllHandler` which will add all the event listeners to expand all rows. For more info, usage and examples: https://react-table.tanstack.com/docs/api/useExpanded#instance-properties

File: src/users/enrollments/EnrollmentsV2.jsx (hunk 8 above)
Comment (Ali-D-Akbar) [reply src/users/enrollments/EnrollmentsV2.jsx]: The biggest pro to use useCallback here is that, it returns memoized data: https://reactjs.org/docs/hooks-reference.html#usecallback

File: src/users/enrollments/EnrollmentsV2.jsx (hunk 9 above)
Comment (Ali-D-Akbar) [reply src/users/enrollments/EnrollmentsV2.jsx]: explained above.

File: src/users/enrollments/EnrollmentsV2.jsx (hunk 10 above)
Comment (Ali-D-Akbar) [reply src/users/enrollments/EnrollmentsV2.jsx]: `getToggleRowExpandedProps` literally expands/collapse the given component by adding props to it.

File: src/users/enrollments/EnrollmentsV2.test.jsx (hunk 12 above)
Comment (Ali-D-Akbar) [reply src/users/enrollments/EnrollmentsV2.test.jsx]: How are we going to check if no action took place? `getToggleAllRowsExpandedProps` is responsible for expanding all rows. This use case doesn't sound necessary to me.

File: src/overrides.scss (hunk 16 above)
Comment (Ali-D-Akbar) [reply src/overrides.scss:22]: It looks good to me. This is after 50% screen size.

Screenshot 2021-09-07 at 1 41 51 PM

File: src/TableV2.jsx (hunk 15 above)
Comment (Ali-D-Akbar) [reply src/TableV2.jsx]: Rewritten them to to `Table`, `UserPage` and `Enrollments` and using v2 while importing.

File: src/components/Table.jsx [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+    sortable: PropTypes.bool,
+  })).isRequired,
+  data: PropTypes.arrayOf(PropTypes.object).isRequired,
+  renderRowSubComponent: PropTypes.func,
+  className: PropTypes.string,
+};
+
+Table.defaultProps = {
+  renderRowSubComponent: null,
+  className: null,
```
Comment (DawoudSheraz) [inline src/components/Table.jsx]: provide a default class name. Also, it can conflict with React's className argument. Rename this prop to something else for sake of having a distinction in custom props.

Comment (DawoudSheraz) [reply src/components/Table.jsx]: If it is possible to avoid having an explicit class name prop, then go ahead with that approach.

File: src/index.jsx (hunk 2 above)
Comment (DawoudSheraz) [reply src/index.jsx]: @Ali-D-Akbar Not addressed.

File: src/users/enrollments/v2/Enrollments.jsx [hunk 18]
```diff
@@ -0,0 +1,268 @@
...
+      )}
+      {enrollmentData
+        ? (
+          <TableV2
+            columns={columns}
+            data={tableData}
+            // We added this as a prop for our table component
+            // Remember, this is not part of the React Table API,
+            // it's merely a rendering option we created for
+            // ourselves
```
Comment (DawoudSheraz) [inline src/users/enrollments/v2/Enrollments.jsx]: This description should ideally go to the Table component itself.

File: src/users/enrollments/v2/Enrollments.test.jsx [hunk 19]
```diff
@@ -0,0 +1,146 @@
...
+    wrapper = mount(<EnrollmentPageWrapper {...props} />);
+    await waitForComponentToPaint(wrapper);
+  });
+
+  afterEach(() => {
+    wrapper.unmount();
+  });
+
+  it('default enrollment data', () => {
+    const collapsible = wrapper.find('h3');
```
Comment (DawoudSheraz) [inline src/users/enrollments/v2/Enrollments.test.jsx]: ah, collapsible?

File: src/users/enrollments/v2/Enrollments.test.jsx [hunk 20]
```diff
@@ -0,0 +1,146 @@
...
+    expect(collapsible.text()).toEqual('Enrollments (2)');
+  });
+
+  it('No Enrollment Data', async () => {
+    jest.spyOn(api, 'getEnrollments').mockImplementationOnce(() => Promise.resolve([]));
+    wrapper = mount(<EnrollmentPageWrapper {...props} />);
+    await waitForComponentToPaint(wrapper);
+    const collapsible = wrapper.find('h3');
```
Comment (DawoudSheraz) [inline src/users/enrollments/v2/Enrollments.test.jsx]: same comment here.

**Review by DawoudSheraz — COMMENTED** [review]
added some other small comments to be addressed before approval

File: src/components/Table.jsx (hunk 17 above)
Comment (Ali-D-Akbar) [reply src/components/Table.jsx]: We can use `styleName` instead so I guess we can go for that.

File: src/index.jsx (hunk 2 above)
Comment (Ali-D-Akbar) [reply src/index.jsx]: Ah, forgot to comment this out.

File: src/users/enrollments/v2/Enrollments.jsx (hunk 18 above)
Comment (Ali-D-Akbar) [reply src/users/enrollments/v2/Enrollments.jsx]: Removing this.

**Review by DawoudSheraz — APPROVED** [review]

**Review by azanbinzahid — APPROVED** [review]
